### ISSUE TITLE
Potential Bug with HYBRID_SHARD and (n, 1) Device Mesh Falling Back to NO_SHARD

### ISSUE THREAD

**Issue by origin-bio** [pytorch/pytorch#154888]
### 🐛 Describe the bug

Hi all,

I ran into unexpected behavior when using FSDP with `ShardingStrategy.HYBRID_SHARD` and a device mesh of shape `(n, 1)`, where `n` is the replicate dimension and `1` is the (non-existent) shard dimension.

As expected, FSDP falls back to using `NO_SHARD` in this case, since sharding over a size-1 dimension is a no-op:

```
UserWarning: FSDP is switching to use `NO_SHARD` instead of ShardingStrategy.HYBRID_SHARD since the world size is 1.
```

However, the fallback seems to set the *world size to 1*, which results in gradients **not** being synchronized across replicated ranks. This causes model parameters to diverge between GPUs.

To verify this, I ran a test on a single node with 2 GPUs and used a replicate dimension of 2. After one training step, I compared model parameters and optimizer states across ranks:

* ✅ With DDP and FSDP + `NO_SHARD`: parameters were consistent.
* ❌ With FSDP + `HYBRID_SHARD` (which falls back to `NO_SHARD`): parameters diverged.

It seems like the fallback isn't preserving the DDP-like behavior as expected. I'm not sure if this is a bug or intentional, but we were hoping to use a single FSDP script to test different mesh configurations (including `(n, 1)`), and this issue breaks that workflow.

### Minimal reproducible code
Run:
```
torchrun --nnodes=1 --nproc_per_node=2 --rdzv_id=12345 --rdzv_backend=c10d --rdzv_endpoint=$MASTER_ADDR:$MASTER_PORT  -m src.training.<train_file>
```
Tweak sharding strategy to see the difference:
`sharding_strategy=ShardingStrategy.HYBRID_SHARD # This is key for the bug, try also with NO_SHARD`
```
import os
import random
import argparse

import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
import torch.distributed as dist
from torch.utils.data import Dataset, DataLoader
from torch.utils.data.distributed import DistributedSampler

from torch.distributed.fsdp import (
    FullyShardedDataParallel as FSDP,
    ShardingStrategy,
)
from torch.distributed.device_mesh import init_device_mesh

import logging

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - RANK %(rank)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)


class ToyModel(nn.Module):
    def __init__(self, input_size=10, hidden_size=20, output_size=5):
        super().__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, output_size)
        self._initialize_weights()

    def _initialize_weights(self):
        # FSDP's sync_module_states=True handles initial parameter consistency.
        for m in self.modules():
            if isinstance(m, nn.Linear):
                nn.init.xavier_uniform_(m.weight)
                if m.bias is not None:
                    nn.init.zeros_(m.bias)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        return x


class RandomDataset(Dataset):
    def __init__(self, num_samples=128, input_size=10, output_size=5, seed=42):
        self.num_samples = num_samples
        rng_state = torch.get_rng_state() # Preserve global RNG state
        torch.manual_seed(seed) # Ensure consistent data generation for the dataset
        self.data = torch.randn(num_samples, input_size)
        self.targets = torch.randn(num_samples, output_size)
        torch.set_rng_state(rng_state) # Restore global RNG state

    def __len__(self):
        return self.num_samples

    def __getitem__(self, idx):
        return {"input": self.data[idx], "target": self.targets[idx], "sample_id": idx}


def setup_distributed(rank, world_size, seed=1629):
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)
    
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)
    
    # try:
    #     # Important for reproducibility, but can error if some ops lack deterministic versions.
    #     torch.use_deterministic_algorithms(True)
    # except RuntimeError as e:
    #     logger.warning(f"Could not enable deterministic algorithms: {e} - proceeding may still show the bug.")

    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    
    # Ensure rank is part of logging format for all loggers
    formatter = logging.Formatter("%(asctime)s - RANK {} - %(levelname)s - %(message)s".format(rank))
    for handler in logging.getLogger().handlers:
        handler.setFormatter(formatter)


def cleanup_distributed():
    dist.destroy_process_group()


def run_bug_repro(rank, world_size, args):
    setup_distributed(rank, world_size, seed=args.seed)
    logger.info(f"Starting bug reproduction on rank {rank} of {world_size}.")

    # Device Mesh: (world_size, 1) for HYBRID_SHARD to mimic DDP-like behavior.
    device_mesh = init_device_mesh("cuda", (world_size, 1), mesh_dim_names=("replicate", "shard"))
    logger.info(f"Device mesh created: {device_mesh}")

    dataset = RandomDataset(num_samples=args.batch_size * world_size * 2, seed=args.seed)
    sampler = DistributedSampler(
        dataset,
        num_replicas=world_size,
        rank=rank,
        shuffle=False, # Critical for reproducibility of data across ranks
        seed=args.seed
    )
    dataloader = DataLoader(dataset, batch_size=args.batch_size, sampler=sampler, num_workers=0)

    model = ToyModel().to(rank)
    
    fsdp_model = FSDP(
        model,
        sharding_strategy=ShardingStrategy.HYBRID_SHARD, # This is key for the bug
        device_mesh=device_mesh,
        use_orig_params=True, # As per user's setup for bug divergence
        sync_module_states=True # Ensures initial params are consistent
    )
    logger.info("Model wrapped with FSDP (HYBRID_SHARD, use_orig_params=True).")

    optimizer = optim.AdamW(fsdp_model.parameters(), lr=1e-3)
    criterion = nn.MSELoss()

    fsdp_model.train()
    
    try:
        batch = next(iter(dataloader))
    except StopIteration:
        logger.error("Dataloader is empty. Increase dataset size or check sampler.")
        cleanup_distributed()
        return

    inputs = batch["input"].to(rank)
    targets = batch["target"].to(rank)
    sample_ids = batch["sample_id"]
    logger.info(f"Rank {rank} processing sample IDs: {sample_ids.tolist()}")

    optimizer.zero_grad()
    outputs = fsdp_model(inputs)
    loss = criterion(outputs, targets)
    loss.backward()

    # Gradient Check: Before optimizer.step() to see raw gradients after backward pass.
    logger.info(f"Rank {rank} performing gradient check...")
    diverged = False
    if world_size > 1:
        for name, param in fsdp_model.named_parameters():
            if param.grad is not None:
                # Get local gradient tensor
                local_grad = param.grad.clone()

                # Create tensor for all-gathered gradients
                gathered_grads = [torch.zeros_like(local_grad) for _ in range(world_size)]
                # All-gather gradients from all ranks
                dist.all_gather(gathered_grads, local_grad)

                # Compare gradients from different ranks (rank 0's grad vs other ranks' grads)
                # Each rank performs this check.
                for i in range(1, world_size):
                    if not torch.allclose(gathered_grads[0], gathered_grads[i], rtol=1e-5, atol=1e-6):
                        logger.warning(
                            f"Rank {rank}: GRADIENT DIVERGENCE DETECTED for param '{name}' between rank 0 and rank {i}."
                        )
                        if rank == 0:
                            logger.warning(f"   Param dtype: {param.dtype}, Grad dtype: {param.grad.dtype}")
                            logger.warning(f"   Rank 0 grad norm: {gathered_grads[0].norm()}")
                            logger.warning(f"   Rank {i} grad norm: {gathered_grads[i].norm()}")
                        diverged = True
                        break # Break from inner loop (checking other ranks for this param) if divergence found
            elif param.requires_grad:
                # This case might indicate an issue if a parameter that should get grads doesn't.
                logger.warning(f"Rank {rank}, Param: {name} requires grad but has no gradient.")
    
    if not diverged and world_size > 1:
        logger.info(f"Rank {rank}: Gradients appear synchronized across ranks after backward pass.")
    elif world_size == 1:
        logger.info(f"Rank {rank}: Running with world_size=1, no gradient synchronization to check.")

    optimizer.step()
    
    logger.info(f"Rank {rank} completed one training step. Loss: {loss.item()}")

    cleanup_distributed()
    logger.info(f"Rank {rank} finished and cleaned up.")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="FSDP Bug Reproduction Script")
    parser.add_argument("--batch_size", type=int, default=4, help="Batch size per GPU")
    parser.add_argument("--seed", type=int, default=42, help="Random seed for setup and data")
    args = parser.parse_args()

    # Assumes script is launched with torchrun, which sets these environment variables.
    # No default values for rank/world_size; torchrun must provide them.
    local_rank = int(os.environ["LOCAL_RANK"])
    rank = int(os.environ["RANK"])
    world_size = int(os.environ["WORLD_SIZE"])
        
    run_bug_repro(rank, world_size, args) 
```

### Versions

PyTorch version: 2.5.1
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.11.11 (main, Dec 11 2024, 16:28:39) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-1023-aws-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB

Nvidia driver version: 535.183.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             48
On-line CPU(s) list:                0-47
Vendor ID:                          GenuineIntel
Model name:                         Intel Xeon Processor (SapphireRapids)
CPU family:                         6
Model:                              143
Thread(s) per core:                 1
Core(s) per socket:                 48
Socket(s):                          1
Stepping:                           4
BogoMIPS:                           4200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk avx512_fp16 arch_capabilities
Virtualization:                     VT-x
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           192 MiB (48 instances)
L3 cache:                           16 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-47
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Unknown: No mitigations
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI Syscall hardening, KVM SW loop
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==2.0.1
[pip3] torch==2.5.1
[pip3] triton==3.1.0
[conda] blas                      1.0                         mkl  
[conda] cuda-cudart               12.4.127                      0    nvidia
[conda] cuda-cupti                12.4.127                      0    nvidia
[conda] cuda-libraries            12.4.1                        0    nvidia
[conda] cuda-nvrtc                12.4.127                      0    nvidia
[conda] cuda-nvtx                 12.4.127                      0    nvidia
[conda] cuda-opencl               12.9.19                       0    nvidia
[conda] cuda-runtime              12.4.1                        0    nvidia
[conda] libcublas                 12.4.5.8                      0    nvidia
[conda] libcufft                  11.2.1.3                      0    nvidia
[conda] libcurand                 10.3.10.19                    0    nvidia
[conda] libcusolver               11.6.1.9                      0    nvidia
[conda] libcusparse               12.3.1.170                    0    nvidia
[conda] libnvjitlink              12.4.127                      0    nvidia
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py311h5eee18b_2  
[conda] mkl_fft                   1.3.11          py311h5eee18b_0  
[conda] mkl_random                1.2.8           py311ha02d727_0  
[conda] numpy                     2.0.1           py311h08b1b3b_1  
[conda] numpy-base                2.0.1           py311hf175353_1  
[conda] pytorch                   2.5.1           py3.11_cuda12.4_cudnn9.1.0_0    pytorch
[conda] pytorch-cuda              12.4                 hc786d27_7    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchtriton               3.1.0                     py311    pytorch

**Comment by weifengpy** [comment]
> It seems like the fallback isn't preserving the DDP-like behavior as expected. I'm not sure if this is a bug or intentional

thanks for reporting! this is a bug. if NO_SHARD all-reduce gradients, HYBRID_SHARD should all-reduce as well.

we might be slow in fixing fsdp1, are you interested in fsdp2? passing 2d device mesh with `fully_shard(mesh=)` is equivalent to fdsp1 HYBRID_SHARD

fsdp2 tutorial: https://docs.pytorch.org/tutorials/intermediate/FSDP_tutorial.html 
2d device mesh example: https://github.com/pytorch/pytorch/blob/3398d1d459f97ef54bb1a6ab66db9f49e2581a72/test/distributed/_composable/fsdp/test_fully_shard_training.py#L1335-L1342

**Comment by origin-bio** [comment]
Hey, thank you for sharing this. Will look into FSDP2. Glad we could help!
